llvm/include/llvm/Transforms/Utils/SYCLModuleSplit.h

```diff
@@ -0,0 +1,224 @@
+struct EntryPointGroup {
+  // Properties an entry point (EP) group
+  struct Properties {
+    // Scope represented by EPs in a group
+    EntryPointsGroupScope Scope = Scope_Global;
+  };
+
+  std::string GroupId;
+  EntryPointSet Functions;
+  Properties Props;
+
+  EntryPointGroup(StringRef GroupId = "") : GroupId(GroupId) {}
+  EntryPointGroup(StringRef GroupId, EntryPointSet Functions)
+      : GroupId(GroupId), Functions(std::move(Functions)) {}
+  EntryPointGroup(StringRef GroupId, EntryPointSet Functions,
+                  const Properties &Props)
+      : GroupId(GroupId), Functions(std::move(Functions)), Props(Props) {}
+
+  void saveNames(std::vector<std::string> &Dest) const;
+  void rebuildFromNames(const std::vector<std::string> &Names, const Module &M);
+  void rebuild(const Module &M);
+};
+
+using EntryPointGroupVec = std::vector<EntryPointGroup>;
+
+// Annotates an llvm::Module with information necessary to perform and track
+// result of device code (llvm::Module instances) splitting:
+// - entry points of the module determined e.g. by a module splitter, as well
+//   as information about entry point origin (e.g. result of a scoped split)
+// - its properties, such as whether it has specialization constants uses
+// It also provides convenience functions for entry point set transformation
+// between llvm::Function object and string representations.
+class ModuleDesc {
+  std::unique_ptr<Module> M;
+  EntryPointGroup EntryPoints;
+
+public:
+  ModuleDesc(std::unique_ptr<Module> M) : M(std::move(M)) {}
+
+  ModuleDesc(std::unique_ptr<Module> M, EntryPointGroup EntryPoints)
+      : M(std::move(M)), EntryPoints(std::move(EntryPoints)) {}
+
+  ModuleDesc(std::unique_ptr<Module> M, const std::vector<std::string> &Names)
+      : M(std::move(M)) {
+    rebuildEntryPoints(Names);
+  }
+
+  const EntryPointSet &entries() const { return EntryPoints.Functions; }
+  const EntryPointGroup &getEntryPointGroup() const { return EntryPoints; }
+  EntryPointSet &entries() { return EntryPoints.Functions; }
+  Module &getModule() { return *M; }
+  const Module &getModule() const { return *M; }
+  std::unique_ptr<Module> releaseModulePtr() { return std::move(M); }
+
+  // Sometimes, during module transformations, some Function objects within the
+  // module are replaced with different Function objects with the same name.
+  // Entry points need to be updated to include the replacement function.
+  // save/rebuild pair of functions is provided to automate this process.
+  void saveEntryPointNames(std::vector<std::string> &Dest) {
+    EntryPoints.saveNames(Dest);
+  }
+
+  void rebuildEntryPoints(const std::vector<std::string> &Names) {
+    EntryPoints.rebuildFromNames(Names, getModule());
+  }
+
+  void rebuildEntryPoints(const Module &M) { EntryPoints.rebuild(M); }
+
+  void rebuildEntryPoints() { EntryPoints.rebuild(*M); }
+
+  // Cleans up module IR - removes dead globals, debug info etc.
+  void cleanup();
+
+  ModuleDesc clone() const;
```

Comment:
I see 'clone' is used in Spec Constant processing here:
https://github.com/asudarsa/llvm-2/blob/dd7446c07ed2ece15fecca65a684aad64a2eb452/llvm/tools/sycl-post-link/sycl-post-link.cpp#L498

May be remove here and add it along with Spec Constant support later?